ludwig/features/sequence_feature.py

```diff
@@ -279,26 +360,146 @@ def predictions(
+                # or deal with it averaging everything
+                # raise ValueError(
+                #     "encoder_output_state has length different than 2 or 4. "
+                #     "Please doublecheck your encoder"
+                # )
+                average_state = average(decoder_input_state)
+                decoder_input_state = [average_state, average_state]
+
+        elif (self.decoder_obj.cell_type == 'lstm' and
+              not isinstance(decoder_input_state, list)):
+            decoder_input_state = [decoder_input_state, decoder_input_state]
+
+        elif (self.decoder_obj.cell_type != 'lstm' and
+              isinstance(decoder_input_state, list)):
+            # here we have a couple options,
+            # either reuse part of the input encoder state,
+            # or just use its output
+            if len(decoder_input_state) == 2:
+                # using h and ignoring c
+                decoder_input_state = decoder_input_state[0]
+            elif len(decoder_input_state) == 4:
+                # using h and ignoring c
+                decoder_input_state + average(
+                    [decoder_input_state[0], decoder_input_state[2]]
+                )
+            else:
+                # no idea how lists of length different than 2 or 4
+                # might have been originated, we can either rise an ValueError
+                # or deal with it averaging everything
+                # raise ValueError(
+                #     "encoder_output_state has length different than 2 or 4. "
+                #     "Please doublecheck your encoder"
+                # )
+                decoder_input_state = average(decoder_input_state)
+            # this returns a [b, h]
+            # decoder_input_state = reduce_sequence(eo, self.reduce_input)
+
+        elif (self.decoder_obj.cell_type != 'lstm' and
+              not isinstance(decoder_input_state, list)):
+            # do nothing, we are good
+            pass
+
+        # at this point decoder_input_state is either a [b,h]
+        # or a list([b,h], [b,h]) if the decoder cell is an lstm
+        # but h may not be the same as the decoder state size,
+        # so we may need to project
+        if isinstance(decoder_input_state, list):
+            for i in range(len(decoder_input_state)):
+                if (decoder_input_state[i].shape[1] !=
+                        self.decoder_obj.state_size):
+                    decoder_input_state[i] = self.decoder_obj.project(
+                        decoder_input_state[i]
+                    )
+        else:
+            if decoder_input_state.shape[1] != self.decoder_obj.state_size:
+                decoder_input_state = self.decoder_obj.project(
+                    decoder_input_state
+                )
+
+        # make sure we are passing back the state tensors in a list
+        if not isinstance(decoder_input_state, list):
+            decoder_input_state = [decoder_input_state]
```

Comment:
why do we need this? at this point, if the cell of the decoder is lstm, we have a `list([b,h], [b,h])` and it it is not, we have just a `[b,h]`, which should be exactly what the decoder needs, wy do we need to wrap it in a list?

Reply:
This code probably came about when I encountered the first of the two TFA issues.  Since I saw from the debugger the inconsistent data structures involved one state tensor wrapped in a list and and the other state tensor not in a list.  I probably decide to make everything a list just to eliminate the possibility I may had an inconsistency in handling the state tensors and inadvertently causing an error.  

Now that TFA seems more solid, I can back this off.  I'll see if I can simplify the code and pass the generator sequence unit test.